Components/mission creation

VitorFCM opened the pull request:
<!--PS: 🇧🇷/🇬🇧 It's ok to write your PR in portuguese 😃-->

## What type of PR is this? (check all applicable)
- [x] ✨ Feature
- [ ] 🐞 Bug Correction
- [ ] 📝 Documentation Update
- [ ] 🚩 Other

# Changes:
<!-- Describe the main changes and the **expected behaviour** (if aplicable) -->
Inicialmente, essa branch foi criada apenas para o desenvolvimento da tabela de criação de variáveis, igual na seguinte foto do figma
<p align="center">
  <img width="430" height="400" src="https://user-images.githubusercontent.com/62776298/129928431-23aa138e-11bb-4c21-beaf-6d7d21d433a8.png">
</p>

Mas com seu desenvolvimento, mais conteúdo não diretamente relacionado com o componente foi adicionado. Esses conteúdos são:

### Variáveis da missão
Para facilitar a criação e utilização das variáveis da missão, duas classes foram propostas: **MissionVariable** e **MissionVariablesList**, ambas localizadas no mesmo arquivo [class_mission_variables.dart](https://github.com/zenitheesc/Zenith-Monitor/blob/components/mission-creation/lib/utils/mixins/class_mission_variables.dart)

### Mission controller
Esse é o bloc que lida com a interação do usuário durante a criação das variáveis da missão. Todos os seus eventos provém do usuário utilizando a tabela, e os estados retornam para a UI com o objeto MissionVariablesList atualizado, ou, com mensagens de erro. Apenas um dos eventos é utilizado para lidar com a mission pipeline, quando o usuário apertar o botão *Iniciar Missão*, queremos que o mission pipeline tenha conhecimento das variáveis da missão atual. Dessa forma, o **Mission Controller** vai disparar um evento do **Mission Pipeline** para atualizar as variáveis da missão  atual. A implementação do **Mission Controller** está localizada em [lib/modules/configuration/bloc/mission_controller/](https://github.com/zenitheesc/Zenith-Monitor/tree/components/mission-creation/lib/modules/configuration/bloc/mission_controller) 

### Mission Pipeline
Pensei que esse pipeline deveria ser o que possui a informação de quais variáveis serão utilizadas missão atual, enquanto que o **Mission Controller** lida apenas com a interação do usuário durante a criação das variáveis. Ele está bem simples, possui apenas um evento para atualizar a informação de seu objeto **MissionVariablesList**. Sua implementação está localizada em [lib/pipelines/mission_pipeline/](https://github.com/zenitheesc/Zenith-Monitor/tree/components/mission-creation/lib/pipelines/mission_pipeline) 

### Página das configurações
Acabei fazendo essa página para prover o bloc **Mission Controller** apenas para a árvore de widgets necessária, que é o widget que possui a tabela da foto. Diferentemente, o bloc  **Mission Pipeline** está sendo instânciado no arquivo **main**, ou seja, fornecido para todo o aplicativo. A implementação da página está localizada em [lib/modules/configuration/screen/configuration.dart](https://github.com/zenitheesc/Zenith-Monitor/blob/components/mission-creation/lib/modules/configuration/screen/configuration.dart)

## Resultado
A seguir um vídeo de como a tabela ficou no meu celular. Por enquanto só é possível adicionar variáveis do tipo inteiro, real e string, mas novas variáveis podem ser incluidas facilmente. No arquivo [lib/constants/variables_types.dart](https://github.com/zenitheesc/Zenith-Monitor/blob/components/mission-creation/lib/constants/variables_types.dart) há listas com as possíveis entradas que o usuário pode fornecer para cada tipo de variável. Também é possível apagar determinada variável dando dois toques sobre o nome da variável que se deseja excluir, por isso que as variáveis começam a sumir em uma parte do vídeo. 

https://user-images.githubusercontent.com/62776298/129949263-a9dbad5b-9146-4238-8c54-a62ba9b17971.mp4   

Closes #25
Closes #64

VitorFCM committed:
Adding basic widget file

igorlourencetti committed:
Add TextField

igorlourencetti committed:
TextFields and button added

VitorFCM committed:
resolving merge conflits

VitorFCM committed:
Creating a possible structure for mission variables

VitorFCM committed:
Few changes in MissionVariables

igorlourencetti committed:
ActionButton and TextFields adaptations

VitorFCM committed:
Merge branch 'feature/variables-table' into components/mission-creation

VitorFCM committed:
just componentizing code for less repetition

VitorFCM committed:
A few changes in UI to look more like figma

VitorFCM committed:
Creating a simple table and implementing Variable Block

VitorFCM committed:
Removing stringToPattern function on variables names, so they can be everything now

VitorFCM committed:
Making variables table look more like figma

VitorFCM committed:
Refactoring the bloc code: making blocBuilder cleaner and changing the state hierarchy

VitorFCM committed:
Adding pubspec.lock changes

VitorFCM committed:
More user interactions

Adding more 2 user user interactions: variable delete and error message when trying to add a variable that already exists

VitorFCM committed:
adding little changes in variables bloc

VitorFCM committed:
Adding start mission button

VitorFCM committed:
Adding mission controller and mission pipeline

The mission controller was the code in modules/mission_detail/bloc. So far, the mission pipeline only has the event to get a new variables_list

VitorFCM committed:
Creating configuration screen

The configuration page was mainly created to provide the mission controller bloc to mission_creation widget, and leave the main file providing only the pipeline bloc

VitorFCM committed:
Merge conflicts

VitorFCM committed:
linting-fix: unnecessary this and new

VitorFCM committed:
linting-fix: added const keywords and others

VitorFCM committed:
Removing unnecessary package

VitorFCM committed:
Changing a MissionVariable property

Instead of maintaining a sequence of values with a list, a single MissionVariable will only store a single value. This has been changed considering that each time a packet is recived in terminal, we only need to add a single value in firebase, and not keep a series of values

jorgemrisco commented on lib/modules/configuration/bloc/mission_controller/variables_bloc.dart:
```diff
@@ -0,0 +1,44 @@
+/// So, it deals with user interation in mission_creation.dart widget
+/// and update the MissionVariableList object in mission pipeline.
+
+class VariablesBloc extends Bloc<VariablesEvent, VariablesState> {
```
Vou fazer uns comentários meio opcionais de nomenclatura, não necessáriamente devem ser feitas modificações e pode ignorar se achar "exagero":

Será que aqui não ficaria mais descritivo o nome se for `MissionVariablesBloc`? Apenas `VariablesBloc` talvez fique muito genérico e pode parecer que o bloc tem um escopo maior do que realmente tem, também segue o mesmo "padrão" de outras nomenclaturas usadas como `MissionBloc` e `MissionVariablesList`

VitorFCM replied:
Concordo também,  o que você acha de ```MissionVariablesController```? Já que temos essa ideia de que **controller** são os blocs que lidam com a UI. Poderíamos manter o padrão de especificar o nome dos blocs entre **pipeline** e **controller**.

jorgemrisco commented on lib/modules/configuration/screen/configuration.dart:
```diff
@@ -0,0 +1,54 @@
+import 'package:zenith_monitor/widgets/mission_creation.dart';
+import 'package:zenith_monitor/widgets/standard_app_bar.dart';
+
+class Configuration extends StatelessWidget {
```
Aqui eu levantaria o mesmo ponto de nomenclatura, talvez `configuration` seja um nome genérico demais e pode passar uma ideia equivocada (ou incompleta) ao leitor, talvez `MissionConfiguration` seja mais específico?
____
A lógica seria que um novo desenvolvedor teria que ler o código para entender o "tipo" de configuração a qual nos estamos referindo, mas talvez podemos ajudar ele já passando essa informação no nome da variável/pasta.

jorgemrisco commented on lib/pipelines/mission_pipeline/mission_bloc.dart:
```diff
@@ -0,0 +1,24 @@
+
+/// This pipeline is still incomplete. So far, its is only being used
+/// to get the variable list from mission_creation widget. Delete these
+/// comments if you are adding more content to this bloc. 
```
Adicionar linha em branco no fim da file

jorgemrisco commented on lib/widgets/mission_creation.dart:
```diff
@@ -0,0 +1,233 @@
+
+  Widget textInputs() {
+    return Table(
+        //defaultVerticalAlignment: TableCellVerticalAlignment.middle,
```
Apagar código comentado
(Se vc não apagar, ninguém vai ter coragem de apagar achando que é importante e aí vamos acumulando)

jorgemrisco commented on lib/modules/configuration/bloc/mission_controller/variables_bloc.dart:
```diff
@@ -0,0 +1,44 @@
+      : super(VariablesInitial(variablesList));
+
+  @override
+  Stream<VariablesState> mapEventToState(VariablesEvent event) async* {
```
`VariablesState` -> `MissionVariablesState` seguindo a lógica anterior;
O equivalente pra análogos tb

jorgemrisco reviewed (approved):
A organização do PR ficou exemplar!

Franreno commented on lib/widgets/mission_creation.dart:
```diff
@@ -0,0 +1,233 @@
+            border: const TableBorder(
+                verticalInside:
+                    BorderSide(width: 1, color: gray, style: BorderStyle.solid),
+                borderRadius: BorderRadius.all(Radius.circular(10))),
```
Eu estava tentando rodar aqui e ta dando problema nessa linha em relação ao borderRadius. Dei uma olhada na documentação mas não achei nada falando sobre o borderRadius dentro do TableBorder. 

![Screenshot from 2021-08-23 09-28-25](https://user-images.githubusercontent.com/67326251/130447118-171128b6-1a73-4fa7-86b3-e7ac08eba5c7.png)

VitorFCM replied:
Boa observação, você poderia dizer qual versão do flutter você está usando? Acho que esse problema é culpa minha, eu esqueci que eu estou na branch **master** e não na **stable** do repositório do flutter. O parâmetro *borderRadius:* foi introduzido recentemente no flutter e acho que nem foi pra branch **stable** ainda. O [PR](https://github.com/flutter/flutter/pull/85946) que adicionaram essa propriedade ainda é muito novo, aparentemente no dia 15/07/2021. O mesmo arquivo da branch **stable** está sem o parâmetro *borderRadius:*, [aqui](https://github.com/flutter/flutter/blob/stable/packages/flutter/lib/src/rendering/table_border.dart).  De qualquer forma, vou voltar para a **stable** para não termos esse problema novamente. Acham que devo alterar o arquivo para ficar compatível ou esperamos que essa alteração chegue na **stable**, já que ela parece ser simples?

Franreno replied:
De fato, minha versão do flutter é a Stable. Eu instalei o flutter pelo Snap e pelo jeito ele pega a versão stable por default.
A minha versão do flutter é a seguinte:
>Flutter is already up to date on channel stable
Flutter 2.2.3 • channel stable • https://github.com/flutter/flutter.git
Framework • revision f4abaa0735 (8 weeks ago) • 2021-07-01 12:46:11 -0700
Engine • revision 241c87ad80
Tools • Dart 2.13.4

Comentando essa linha o app continua a funcionar normalmente, conforme mostra a imagem abaixo. Então eu acho que deixar essa linha comentada por enquanto e abrir um issue pra 'descomentar' ela quando essa funcionalidade for inserida na versão stable possa ser uma solução temporária.
![a](https://user-images.githubusercontent.com/67326251/130489384-9bf02531-7da7-4c62-b84e-6ee240b2c9dc.jpeg)

VitorFCM replied:
Boa Fran, o *borderRadius* estava redundante, pois as rows já definem um raio também. Acabei retirando o parâmetro  *borderRadius* da **TableBorder**, ai a gente não precisa se preocupar.

mairacanal commented on lib/pipelines/mission_pipeline/mission_bloc.dart:
```diff
@@ -0,0 +1,24 @@
+part 'mission_state.dart';
+part 'mission_event.dart';
+
+class MissionBloc extends Bloc<MissionEvent, MissionState> {
```
Não seria melhor se as pipelines ficassem no core?

mairacanal commented on lib/utils/mixins/class_mission_variables.dart:
```diff
@@ -0,0 +1,103 @@
+  }
+}
+
+class VariableAlreadyExistsException implements Exception {
```
Será que valeria a pena criar um arquivo para as Exceptions?

mairacanal commented on lib/widgets/mission_creation.dart:
```diff
@@ -0,0 +1,233 @@
+import 'package:flutter_bloc/flutter_bloc.dart';
+import 'package:zenith_monitor/constants/colors_constants.dart';
+import 'package:flutter/material.dart';
```
Isso é TOC meu, mas será que não fica mais organizado se colocar os pacotes do flutter juntos antes dos nossos?

VitorFCM committed:
Removing unnecessary line

VitorFCM committed:
Removing unnecessary comment

VitorFCM committed:
Wraping the message error with a container to avoid bottom overflowed error

VitorFCM committed:
Adding basic modifications to use mission name

VitorFCM committed:
Change in the configuration of mission variables files

VitorFCM committed:
Variables bloc dealing with mission name errors

The mission name variable has became a property in the MissionVariableList class

VitorFCM committed:
Renaming VariablesBloc to MissionVariablesBloc

VitorFCM committed:
Renaming configuration screen file and mission_controller states and events files

VitorFCM committed:
Just refactoring files to a better pattern

Changing the order of package imports and adding a new line at the end of a file

VitorFCM committed:
Moving pipelines folders to core

VitorFCM committed:
Removing pipelines folder from lib root
